File: 1501-circle-and-rectangle-overlapping/circle-and-rectangle-overlapping.py
class Solution(object):
    def checkOverlap(self, radius, xCenter, yCenter, x1, y1, x2, y2):
        
        # Clamp function
        def clamp(val, low, high):
            return max(low, min(val, high))
        
        # Closest point inside rectangle
        closestX = clamp(xCenter, x1, x2)
        closestY = clamp(yCenter, y1, y2)
        
        # Distance squared
        dx = xCenter - closestX
        dy = yCenter - closestY
        
        return dx * dx + dy * dy <= radius * radius

File: 1501-circle-and-rectangle-overlapping/circle-and-rectangle-overlapping.dart
class Solution {
  bool checkOverlap(int radius, int xCenter, int yCenter,
      int x1, int y1, int x2, int y2) {
    
    int closestX = xCenter.clamp(x1, x2);
    int closestY = yCenter.clamp(y1, y2);

    int dx = xCenter - closestX;
    int dy = yCenter - closestY;

    return dx * dx + dy * dy <= radius * radius;
  }
}